File: Classes/Base/NSObject+LWTheme.m
//
//  NSObject+LWTheme.m
//  LWTheme
//
//  Created by lalawue on 2019/12/8.
//  Copyright © 2019 lalawue. All rights reserved.
//

#import "NSObject+LWTheme.h"
#import "LWThemeManager_Private.h"
#import <objc/runtime.h>

static void const *kNSObjectLWThemeColorSetterKey = &kNSObjectLWThemeColorSetterKey;

@implementation NSObject (LWTheme)

- (void)lw_storeValue:(id)value forSetter:(SEL)setter
{
    if (!setter) {
        return;
    }
    NSString *setterKey = NSStringFromSelector(setter);

    NSMutableDictionary *dict = objc_getAssociatedObject(self, &kNSObjectLWThemeColorSetterKey);
    if (!dict) {
        dict = [NSMutableDictionary dictionaryWithCapacity:2];
        objc_setAssociatedObject(self, &kNSObjectLWThemeColorSetterKey, dict, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
    }

    if (value) {
        [dict setObject:value forKey:setterKey];
    } else {
        [dict removeObjectForKey:setterKey];
    }
    
    if (dict.count) {
        [[LWThemeManager sharedInstance] addThemeObject:self];
    } else {
        [[LWThemeManager sharedInstance] removeThemeObject:self];
    }
}

- (id)lw_valueForSetter:(SEL)setter
{
    if (setter) {
        NSString *setterKey = NSStringFromSelector(setter);
        NSMutableDictionary *dict = objc_getAssociatedObject(self, &kNSObjectLWThemeColorSetterKey);
        return [dict objectForKey:setterKey];
    } else {
        return @"";
    }
}

- (void)lw_callSetterWithValue
{
    NSMutableDictionary *dict = objc_getAssociatedObject(self, &kNSObjectLWThemeColorSetterKey);
    for (NSString *setterKey in dict.allKeys) {
        SEL setter = NSSelectorFromString(setterKey);
        if ([self respondsToSelector:setter]) {
            id value = [dict objectForKey:setterKey];
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
            [self performSelector:setter withObject:value];
#pragma clang diagnostic pop
        }
    }
}

- (UIColor *)lw_colorForPattern:(NSString *)colorPattern
{
    return [[LWThemeManager sharedInstance] colorForPattern:colorPattern];
}

+ (UIColor *)lw_colorForPattern:(NSString *)colorPattern
{
    return [[LWThemeManager sharedInstance] colorForPattern:colorPattern];
}

- (UIFont *)lw_fontForPattern:(NSString *)fontPattern
{
    return [[LWThemeManager sharedInstance] fontForPattern:fontPattern];
}

- (UIImage *)lw_imageNamed:(NSString *)imagePattern
{
    return [[LWThemeManager sharedInstance] imageNamed:imagePattern];
}

@end

File: Classes/Base/NSObject+LWTheme.swift
//
//  NSObject+LWTheme.swift
//  LWThemeSwift
//
//  Created by lalawue on 2019/12/17.
//  Copyright © 2019 lalawue. All rights reserved.
//

import UIKit

protocol LWThemeObjectNotify : NSObject {
    func lw_callSetterWithValue(); ///< call when theme changed
}

extension NSObject : LWThemeObjectNotify {
    
    fileprivate struct AssociatedKey {
        static var address : UInt8 = 0
    }
    
    func lw_storeValue(_ value : Any?, setter : Selector) -> Void {
        let setterKey = NSStringFromSelector(setter)
        var dict = objc_getAssociatedObject(self, &AssociatedKey.address) as? NSMutableDictionary
        if (dict == nil) {
            dict = NSMutableDictionary(capacity: 2);
            objc_setAssociatedObject(self, &AssociatedKey.address, dict, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        }
        if (value != nil) {
            dict!.setObject(value as Any, forKey: setterKey as NSCopying);
        } else {
            dict!.removeObject(forKey: setterKey)
        }
        
        if (dict!.count > 0) {
            LWThemeManager.sharedInstance.addThemeObject(self)
        } else {
            LWThemeManager.sharedInstance.removeThemeObject(self)
        }
    }
    
    func lw_valueForSetter(_ setter : Selector) -> Any? {
        let dict = objc_getAssociatedObject(self, &AssociatedKey.address) as? NSMutableDictionary
        if ((dict != nil) && (dict!.count > 0)) {
            let setterKey = NSStringFromSelector(setter)
            return dict![setterKey]
        } else {
            return ""
        }
    }

    func lw_colorForPattern(_ colorPattern : String?) -> UIColor? {
        return LWThemeManager.sharedInstance.colorForPattern(colorPattern)
    }
    
    func lw_fontForPattern(_ fontPattern : String?) -> UIFont? {
        return LWThemeManager.sharedInstance.fontForPattern(fontPattern)
    }
    
    func lw_imageNamed(_ name : String?) -> UIImage? {
        return LWThemeManager.sharedInstance.imageNamed(name)
    }
    
    //MARK: Apply
    
    @objc func lw_callSetterWithValue() {
        let dict = objc_getAssociatedObject(self, &AssociatedKey.address) as? NSMutableDictionary
        if (dict != nil) {
            for setterKey in dict!.allKeys {
                let selector = NSSelectorFromString(setterKey as! String)
                if (responds(to: selector)) {
                    perform(selector, with: dict![setterKey])
                }
            }
        }
    }
}
